### PR TITLE
Adds isomorphic/server-side React application-shell rendering

### PR DESCRIPTION
**Pull request by addyosmani**
Demo URL: https://react-hn.appspot.com

**tl;dr: we can now server-side render the UI, which also enables it to be visible with or without JavaScript enabled**

This set of changes formalises the work done in #27 (huge thanks to @jackfranklin and @psimyn) for helping us get it started. The end product can now be trivially deployed to Google App Engine.

@jackfranklin: some of the specific changes I added to your branch were bringing back SSR-friendly hostname parsing, the `componentDidMount` fix, slightly tighter equality checks and updates to the build pipeline which enable us to deploy to GAE. 

cc @insin @NekR

### CODE REVIEW

**Comment by NekR** [comment]
I am not really React-guy, but if no one will be able to review it, then I'll do review.

**Comment by addyosmani** [comment]
Thanks! Any help at all with reviews would be 💕

**Comment by jackfranklin** [comment]
Awesome - I'll take a look through quickly if that's useful :)

**Comment by addyosmani** [comment]
@jackfranklin That would be super useful :) @insin would you like for us to wait on your final call after other reviews are done before merging this in if we get go aheads?

File: src/Stories.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ var Stories = React.createClass({
     }
   },
 
-  componentWillMount() {
+  componentDidMount() {
```
Comment (jackfranklin) [inline src/Stories.js]: more out of interest, but why is this required? Is this what gets around that weird React error I was getting?

**Comment by jackfranklin** [comment]
Looks great!

File: src/Stories.js (hunk 1 above)
Comment (insin) [inline src/Stories.js]: `componentDidMount()` only gets called on the client

File: app.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+runtime: nodejs
```
Comment (insin) [inline app.yaml]: Does this come with any predefined environment variables?

If not, [setting `NODE_ENV` to `production`](http://expressjs.com/en/advanced/best-practice-performance.html#set-nodeenv-to-production) will make Express cache the view template

**Comment by insin** [comment]
You could use something like this [`<Title/>`](https://github.com/ryanflorence/react-title-component) component for isomorphic title handling, but that's pretty minor and easy to add after this PR.

File: src/Stories.js (hunk 1 above)
Comment (jackfranklin) [inline src/Stories.js]: Ah, of course.

File: package.json [hunk 3]
```diff
@@ -9,26 +9,33 @@
     "url": "http://github.com/insin/react-hn.git"
   },
   "scripts": {
-    "build": "npm run lint && cp node_modules/sw-toolbox/sw-toolbox.js public/sw-toolbox.js && nwb build && npm run precache",
-    "lint": "eslint src",
-    "lint:fix": "eslint --fix .",
-    "start": "nwb serve",
-    "precache": "sw-precache --root=public --config=sw-precache-config.json"
+    "build": "npm run lint && cp node_modules/sw-toolbox/sw-toolbox.js public/sw-toolbox.js && ./node_modules/nwb/lib/bin/nwb.js build && npm run precache",
```
Comment (tracker1) [inline package.json]: can replace `./node_modules/nwb/lib/bin/nwb.js` with `./node_modules/.bin/nwb`

File: src/utils/storage.js [hunk 4]
```diff
@@ -1,9 +1,14 @@
 module.exports = {
   get(key, defaultValue) {
-    var value = window.localStorage[key]
-    return (typeof value != 'undefined' ? value : defaultValue)
+    if (typeof window !== 'undefined') {
```
Comment (tracker1) [inline src/utils/storage.js]: if (typeof window === 'undefined') return defaultValue;

File: app.yaml (hunk 2 above)
Comment (addyosmani) [inline app.yaml]: You're right. It looks like GAE does support environment variables: https://cloud.google.com/appengine/docs/flexible/nodejs/configuring-your-app-with-app-yaml#Node.js_app_yaml_Defining_environment_variables. I'll update.

File: package.json (hunk 3 above)
Comment (addyosmani) [inline package.json]: Good catch. Will update.

File: src/utils/storage.js (hunk 4 above)
Comment (addyosmani) [inline src/utils/storage.js]: Updating to reflect feedback in just a sec.

**Comment by addyosmani** [comment]
@insin @tracker1 Attempted to address feedback. PTAL if you get a chance :)

**Comment by insin** [comment]
LGTM!
